### PR TITLE
Replace WebSocket assertions with RuntimeError

### LINKED ISSUE #1471: Where assert statements are guarding against invalid ASGI messaging, use RuntimeError instead.
### Checklist

- [X] There are no similar issues or pull requests for this yet.
- [X] I discussed this idea on the [community chat](https://gitter.im/encode/community) and feedback is positive.

### Is your feature related to a problem? Please describe.

There are `assert` statements in the source code which raise a vague and hard to debug `AssertionError`. For example on [this line](https://github.com/encode/starlette/blob/f12c92a21500d484b3d48f965bb605c1bbe193bc/starlette/websockets.py#L58).

If some kind of exception (for example something along the lines of: `WebSocketMessageType`) were raised it would make debugging a lot clearer. I spent a lot more time than I should have just working out where exactly this `AssertionError` was coming from and what the root cause was.

### Describe the solution you would like.

This is by no means the right solution but at least it's an idea of the kind of thing that might help:

```python
class WebSocketMessageType(Exception):
    pass

class WebSocket(HTTPConnection):
    ...
    async def send(self, message: Message) -> None:
        """
        Send ASGI websocket messages, ensuring valid state transitions.
        """
        if self.application_state == WebSocketState.CONNECTING:
            message_type = message["type"]
            if message_type not in {"websocket.accept", "websocket.close"}:
                raise WebSocketMessageType("expected message_type to be websocket.accept or websocket.close")
            if message_type == "websocket.close":
                self.application_state = WebSocketState.DISCONNECTED
            else:
                self.application_state = WebSocketState.CONNECTED
            await self._send(message)
        elif self.application_state == WebSocketState.CONNECTED:
            message_type = message["type"]
            if message_type not in {"websocket.send", "websocket.close"}:
                raise WebSocketMessageType("expected message_type to be websocket.send or websocket.close")
            if message_type == "websocket.close":
                self.application_state = WebSocketState.DISCONNECTED
            await self._send(message)
        else:
            raise RuntimeError('Cannot call "send" once a close message has been sent.')
```

### Describe alternatives you considered

_No response_

### Additional context

The error I was seeing:

```
ERROR    root:a_file.py:31 {'message': 'Job processing failed', 'job': <Job coro=<<coroutine object a_class.a_method at 0x7f6d7a7c1ec0>>>, 'exception': AssertionError()}
NoneType: None
```

And this would be it with a `raise` statement: admittedly there is still no mention of `starlette` so a user would still have to diagnose that as the root cause.

```
ERROR    root:a_file.py:31 {'message': 'Job processing failed', 'job': <Job coro=<<coroutine object a_class.a_method at 0x7fb99c2ed940>>>, 'exception': WebSocketMessageType('expected message_type to be websocket.accept or websocket.close')}
NoneType: None
```

Also, I have no idea where that `NoneType: None` is coming from or what that means.

### PR DESCRIPTION
**Pull request by aminalaee**
Fixes #1471.

Switching public facing WebSocket assertions to RuntimeError.

### CODE REVIEW

File: tests/test_websockets.py [hunk 1]
```diff
@@ -422,3 +422,135 @@ async def asgi(receive, send):
...
+        with client.websocket_connect("/"):
+            pass  # pragma: nocover
+
+
+def test_receive_before_accept(test_client_factory):
+    def app(scope):
+        async def asgi(receive, send):
+            websocket = WebSocket(scope, receive=receive, send=send)
+            await websocket.accept()
+            websocket.client_state = WebSocketState.CONNECTING
```
Comment (aminalaee) [inline tests/test_websockets.py:533]: I couldn't test this without modifying the client state.

**Review by lovelydinosaur — COMMENTED** [review]
Looks great yup! I've got a small suggestion on the text for the four `message_type` cases, which I'll address inline.

File: starlette/websockets.py [hunk 2]
```diff
@@ -34,13 +34,20 @@ async def receive(self) -> Message:
         if self.client_state == WebSocketState.CONNECTING:
             message = await self._receive()
             message_type = message["type"]
-            assert message_type == "websocket.connect"
+            if message_type != "websocket.connect":
+                raise RuntimeError(
+                    'WebSocket is not connected. Need to call "accept" first.'
+                )
```
Comment (lovelydinosaur) [inline starlette/websockets.py:41]: What do we think about this kind of style for the `message_type` case?...

`f'Expected ASGI message "websocket.connect", but got {message_type!r}'`

File: starlette/websockets.py [hunk 3]
```diff
@@ -34,13 +34,20 @@ async def receive(self) -> Message:
...
             return message
         elif self.client_state == WebSocketState.CONNECTED:
             message = await self._receive()
             message_type = message["type"]
-            assert message_type in {"websocket.receive", "websocket.disconnect"}
+            if message_type not in {"websocket.receive", "websocket.disconnect"}:
+                raise RuntimeError(
+                    "Websocket is connected."
+                    'Message type should be either "receive" or "disconnect".'
+                )
```
Comment (lovelydinosaur) [inline starlette/websockets.py:51]: Also here...

File: starlette/websockets.py [hunk 4]
```diff
@@ -55,15 +62,22 @@ async def send(self, message: Message) -> None:
         """
         if self.application_state == WebSocketState.CONNECTING:
             message_type = message["type"]
-            assert message_type in {"websocket.accept", "websocket.close"}
+            if message_type not in {"websocket.accept", "websocket.close"}:
+                raise RuntimeError(
+                    'WebSocket is not connected. Need to call "accept" first.'
+                )
```
Comment (lovelydinosaur) [inline starlette/websockets.py:70]: ...here...

File: starlette/websockets.py [hunk 5]
```diff
@@ -55,15 +62,22 @@ async def send(self, message: Message) -> None:
...
                 self.application_state = WebSocketState.CONNECTED
             await self._send(message)
         elif self.application_state == WebSocketState.CONNECTED:
             message_type = message["type"]
-            assert message_type in {"websocket.send", "websocket.close"}
+            if message_type not in {"websocket.send", "websocket.close"}:
+                raise RuntimeError(
+                    "Websocket is connected."
+                    'Message type should be either "send" or "close".'
+                )
```
Comment (lovelydinosaur) [inline starlette/websockets.py:82]: ...and here.

File: starlette/websockets.py [hunk 6]
```diff
@@ -34,13 +34,21 @@ async def receive(self) -> Message:
         if self.client_state == WebSocketState.CONNECTING:
             message = await self._receive()
             message_type = message["type"]
-            assert message_type == "websocket.connect"
+            if message_type != "websocket.connect":
+                raise RuntimeError(
+                    'Expected ASGI message "websocket.connect",'
```
Comment (Kludex) [inline starlette/websockets.py]:
```suggestion
                    'Expected ASGI message "websocket.connect", '
```

File: starlette/websockets.py [hunk 7]
```diff
@@ -34,13 +34,21 @@ async def receive(self) -> Message:
...
+                )
             self.client_state = WebSocketState.CONNECTED
             return message
         elif self.client_state == WebSocketState.CONNECTED:
             message = await self._receive()
             message_type = message["type"]
-            assert message_type in {"websocket.receive", "websocket.disconnect"}
+            if message_type not in {"websocket.receive", "websocket.disconnect"}:
+                raise RuntimeError(
+                    'Expected ASGI message "websocket.receive" or'
```
Comment (Kludex) [inline starlette/websockets.py]:
```suggestion
                    'Expected ASGI message "websocket.receive" or '
```

File: starlette/websockets.py [hunk 8]
```diff
@@ -55,15 +63,23 @@ async def send(self, message: Message) -> None:
         """
         if self.application_state == WebSocketState.CONNECTING:
             message_type = message["type"]
-            assert message_type in {"websocket.accept", "websocket.close"}
+            if message_type not in {"websocket.accept", "websocket.close"}:
+                raise RuntimeError(
+                    'Expected ASGI message "websocket.connect",'
```
Comment (Kludex) [inline starlette/websockets.py]:
```suggestion
                    'Expected ASGI message "websocket.connect", '
```

File: starlette/websockets.py [hunk 9]
```diff
@@ -55,15 +63,23 @@ async def send(self, message: Message) -> None:
...
                 self.application_state = WebSocketState.DISCONNECTED
             else:
                 self.application_state = WebSocketState.CONNECTED
             await self._send(message)
         elif self.application_state == WebSocketState.CONNECTED:
             message_type = message["type"]
-            assert message_type in {"websocket.send", "websocket.close"}
+            if message_type not in {"websocket.send", "websocket.close"}:
+                raise RuntimeError(
+                    'Expected ASGI message "websocket.send" or "websocket.close",'
```
Comment (Kludex) [inline starlette/websockets.py]:
```suggestion
                    'Expected ASGI message "websocket.send" or "websocket.close", '
```

**Review by lovelydinosaur — APPROVED** [review]

**Comment by dave-fitzgerald-bc** [comment]
What version of `starlette` is this likely to be included in?

**Comment by aminalaee** [comment]
@dave-fitzgerald-bc It will be available in `0.19.0`. You can track it here: https://github.com/encode/starlette/pull/1439

**Pull request merged**